cli/src/lib/dealClient.ts

```diff
@@ -685,3 +685,64 @@ export async function batchRead<T>(rpcReadCalls: Array<() => Promise<T>>) {
 
   return rpcResults;
 }
+
+export async function guessTxSizeAndSign<
+  T,
+  A extends Array<unknown> = Array<unknown>,
+  R = unknown,
+  S extends Exclude<StateMutability, "view"> = "payable",
+>({
+  sliceValuesToRegister,
+  sliceIndex: sliceIndexArg,
+  getArgs,
+  getTitle,
+  ...signArgs
+}: {
+  sliceValuesToRegister: (sliceIndex: number) => Array<T>;
+  sliceIndex: number;
+} & Omit<SignArgs<A, R, S>, "args" | "title"> & {
+    getArgs: (
+      valuesToRegister: T[],
+    ) => Parameters<TypedContractMethod<A, R, S>>;
+    getTitle: (arg: { valuesToRegister: T[]; sliceCount: number }) => string;
+  }) {
+  let valuesToRegister;
+  let sliceIndex = sliceIndexArg;
+  let isValidTx = false;
+  const { providerOrWallet } = await getDealClient();
+  const address = await getSignerAddress();
+
+  do {
+    valuesToRegister = sliceValuesToRegister(sliceIndex);
+
+    try {
+      const populatedTx = await populateTx(
+        signArgs.method,
+        ...getArgs(valuesToRegister),
+      ).populate();
+
+      populatedTx.from = address;
+      await providerOrWallet.estimateGas(populatedTx);
+
+      isValidTx = true;
+    } catch (e) {
+      sliceIndex = Math.floor(sliceIndex / 2);
```

Comment:
maybe better `Math.ceil`? if we check bigger half, smaller half more probably will be ok

Reply:
if I am not mistaken the number of transactions will be the same in either case

Reply:
but if we floored passes, second part may fail if it has more count (+1). In case of ceil, second part may have same or -1 count

Reply:
It will not fail. Rest of the parts are checked exactly the same way (there can be more then 2 if a lot of CUs it can be a lot of txs)

Reply:
ok